src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,792 @@
+            }
+
+            private bool TryProcessFromClause(FromClauseSyntax fromClause, ProcessingState processingState)
+            {
+                var queryClauseInfo = _semanticModel.GetQueryClauseInfo(fromClause, _cancellationToken);
+                if (queryClauseInfo.OperationInfo.Symbol == null || // First FromClause may have no OperationInfo symbols.
+                    (queryClauseInfo.OperationInfo.Symbol is IMethodSymbol methodSymbol && methodSymbol.Parameters.Length == 2))
+                {
+                    processingState.AddVariableName(fromClause.Identifier.ValueText);
+                    processingState.AddStatementProcessingMethod(
+                        s => SyntaxFactory.ForEachStatement(
+                                    fromClause.Type ?? VarNameIdentifier,
+                                    fromClause.Identifier,
+                                    fromClause.Expression.WithoutTrivia(), WrapWithBlock(s)));
+                    return true;
+                }
+
+                return false;
+            }
+
+            private bool TryProcessLetClause(LetClauseSyntax letClause, ProcessingState processingState)
+            {
+                var queryClauseInfo = _semanticModel.GetQueryClauseInfo(letClause, _cancellationToken);
+                if (queryClauseInfo.OperationInfo.Symbol is IMethodSymbol methodSymbol &&
+                    methodSymbol.Parameters.Length == 1)
+                {
+                    processingState.AddVariableName(letClause.Identifier.ValueText);
+                    processingState.AddStatementProcessingMethod(
+                        s => AddToBlockTop(
+                                SyntaxFactory.LocalDeclarationStatement(
+                                    SyntaxFactory.VariableDeclaration(
+                                        VarNameIdentifier,
+                                        SyntaxFactory.SingletonSeparatedList(
+                                            SyntaxFactory.VariableDeclarator(
+                                                letClause.Identifier,
+                                                argumentList: null,
+                                                SyntaxFactory.EqualsValueClause(letClause.Expression))))),
+                                s));
+
+                    return true;
+                }
+
+                return false;
+            }
+
+            private bool TryProcessJoinClause(JoinClauseSyntax joinClause, ProcessingState processingState)
+            {
+                var queryClauseInfo = _semanticModel.GetQueryClauseInfo(joinClause, _cancellationToken);
+                if (queryClauseInfo.OperationInfo.Symbol is IMethodSymbol methodSymbol &&
+                    methodSymbol.Parameters.Length == 4)
+                {
+                    processingState.AddVariableName(joinClause.Identifier.ValueText);
+                    if (joinClause.Into != null)
```

Comment:
> if (joinClause.Into != null) [](start = 19, length = 29)

A ```join``` clause with ```into``` is getting mapped to a GroupJoin  operator, the translation looks very suspicious for this case. Is this case covered by tests or even reachable?

Reply:
We bail out on 'group' now. Do you see an option to support it? 

---
In reply to: [177204456](https://github.com/dotnet/roslyn/pull/25362#discussion_r177204456) [](ancestors = 177204456)

Reply:
> We bail out on 'group' now. Do you see an option to support it? 

Even if it is possible, that won't be trivial. I think we don't need to support it right from the start.

---
In reply to: [177578113](https://github.com/dotnet/roslyn/pull/25362#discussion_r177578113) [](ancestors = 177578113,177204456)